apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,441 @@
+type StagedUpdateStrategy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Stage configurations for each update stage.
+	// +kubebuilder:validation:MaxItems=127
+	// +kubebuilder:validation:Required
+	Stages []StageConfig `json:"stages"`
+}
+
+// StagedUpdateStrategyList contains a list of StagedUpdateStrategy.
+// +kubebuilder:resource:scope="Namespaced"
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type StagedUpdateStrategyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []StagedUpdateStrategy `json:"items"`
+}
+
+// StageConfig describes a single update stage group configuration.
+// The clusters in each stage are updated sequentially for now.
+// We will stop the update if any of the updates fail.
+type StageConfig struct {
+	// The name of the stage. This MUST be unique within the same StagedUpdateStrategy.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// LabelSelector is a label query over all the joined member clusters. Clusters matching the query are selected
+	// for this stage. There cannot be overlapping clusters between stages when the stagedUpdateRun is created.
+	// If the label selector is absent, all the selected clusters are selected.
+	// +kubebuilder:validation:Optional
+	LabelSelector *metav1.LabelSelector `json:"labelSelector"`
+
+	// The label key used to sort the selected clusters.
+	// The clusters within the stage are updated sequentially following the rule below:
+	//   - primary: Ascending order based on the value of the label key, interpreted as integers if present.
+	//   - secondary: Ascending order based on the name of the cluster if the label key is absent.
+	// +kubebuilder:validation:Optional
+	SortingLabelKey *string `json:"sortingLabelKey,omitempty"`
+
+	// The collection of tasks that each stage needs to complete successfully before moving to the next stage.
+	// Each task is executed in parallel and they can not be more than one task of the same type.
+	// +kubebuilder:validation:Optional
+	AfterStageTasks []AfterStageTask `json:"afterStageTasks,omitempty"`
+}
+
+// AfterStageTask is the collection of post stage tasks that ALL need to be completed before we can move to the next stage.
+type AfterStageTask struct {
+	// The type of the after stage task.
+	// +kubebuilder:validation:Enum=TimedWait;Approval
+	// +kubebuilder:validation:Required
+	Type AfterStageTaskType `json:"type"`
+
+	// The prefix of the name of the StageApprovalRequest we need to create to get the approval for the next stage.
+	// We create the full name of the approval request by concatenating the prefix to the stagedUpdateRunName and the stageName.
+	// For example, if the approvalRequestPrefix is "myApproval", the name of the approval request for the stage "stage1"
+	// in the update run "updateRun1" is "myApproval-updateRun1-stage1".
```

Comment:
Shall we add a timestamp based or a random postfix to avoid name collisions?

Since spec is immutable, customers might need to delete and recreate the same upddaterun when the updaterun fails, which will lead to the same approval request name.